src/gmpints.c

```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
 #endif
 }
 
+/**************************************************************************
+**
+** The following functions a GAP integer into an Int, UInt, Int8 or UInt8
+** if it is in range, even if it is too big to be a small integer
+*/
+Int Int_ObjInt(Obj i) {
+    UInt sign = 0;
+    if (IS_INTOBJ(i))
+        return INT_INTOBJ(i);
+    // must be a single limb
+    assert( sizeof(mp_limb_t) == sizeof(Int));
+    if (TNUM_BAG(i) == T_INTPOS)
+        sign = 0;
+    else if(TNUM_BAG(i) == T_INTNEG)
+        sign = 1;
+    else
+        ErrorMayQuit("Conversion error, expecting an integer, not a %s",(Int)TNAM_OBJ(i),0);
+    if (SIZE_BAG(i) != sizeof(mp_limb_t))
+        ErrorMayQuit("Conversion error, integer too large", 0L, 0L);
+    UInt val = ADDR_INT(i)[0];
+#ifdef SYS_IS_64_BIT
+    if ((!sign && (val > INT64_MAX)) ||
+        (sign && (val > (UInt)INT64_MIN))) // works because of twos complement arithmetic
+#else
+    if ((!sign && (val > INT32_MAX)) ||
+        (sign && (val > (UInt)INT32_MIN))) // works because of twos complement arithmetic
+#endif
+        ErrorMayQuit("Conversion error, integer too large", 0L, 0L);
+    return sign ? -(Int)val : (Int)val;
+}
+
+UInt UInt_ObjInt(Obj i)
+{
+    if (IS_INTOBJ(i)) {
+        Int val = INT_INTOBJ(i);
+        if (val < 0)
+            ErrorMayQuit("Conversion: negative integer into unsigned type",0,0);
```

Comment:
Instead of having this error twice, you could just start the function like so:
```C
UInt UInt_ObjInt(Obj i)
{
  if (IS_NEGATIVE(i))
    ErrorMayQuit("Conversion: negative integer into unsigned type",0,0);
  if (IS_INTOBJ(i))
    return INT_INTOBJ(i);
  if (TNUM_BAG(i) != T_INTPOS)
    .. 
```

Reply:
Of course with PR #1903, `IS_NEGATIVE` is renamed to `IS_NEG_INT` -- it's a pity these two PRs are clashing now. 

Since I am probably less annoyed by adapting, I guess we can merge this PR first and then I'll adapt PR #1903. *shrug*